gcp_variant_transforms/bq_to_vcf.py

```diff
@@ -56,19 +65,78 @@ def run(argv=None):
   logging.info('Command: %s', ' '.join(argv or sys.argv))
   known_args, pipeline_args = vcf_to_bq_common.parse_args(argv,
                                                           _COMMAND_LINE_OPTIONS)
+  (project, dataset, table) = bigquery_util.parse_table_reference(
+      known_args.input_table)
   bq_source = bigquery.BigQuerySource(
       query=_BASE_QUERY_TEMPLATE.format(
-          INPUT_TABLE='.'.join(bigquery_util.parse_table_reference(
-              known_args.input_table))),
+          INPUT_TABLE='.'.join([project, dataset, table])),
       validate=True,
       use_standard_sql=True)
 
   options = pipeline_options.PipelineOptions(pipeline_args)
+
+  blob_prefix = '_sharded_vcf_files_{}'.format(
+      datetime.now().strftime('%Y%m%d_%H%M%S'))
+  file_path_prefix = '/'.join(['gs:/', known_args.bucket_name, blob_prefix])
   with beam.Pipeline(options=options) as p:
     _ = (p | 'ReadFromBigQuery ' >> beam.io.Read(bq_source)
          | bigquery_to_variant.BigQueryToVariant()
          | densify_variants.DensifyVariants()
-         | vcfio.WriteToVcf(known_args.output_file))
+         | 'PairVariantWithKey' >> beam.Map(_pair_variant_with_key,
+                                            _NUMBER_OF_VARIANTS_PER_SHARD)
+         | 'GroupVariantsByKey' >> beam.GroupByKey()
+         | vcfio.WriteVcfDataLines(file_path_prefix))
+
+  client = storage.Client(project)
+  bucket = client.get_bucket(known_args.bucket_name)
+  composed_vcf_data_file = _compose_multiple_files_to_one(bucket, blob_prefix)
+  output_file_blob = bucket.blob(known_args.output_file)
+  output_file_blob.content_type = 'text/plain'
+  output_file_blob.rewrite(composed_vcf_data_file)
+  bucket.delete_blobs(bucket.list_blobs(prefix=blob_prefix))
+
+
+def _pair_variant_with_key(variant, number_of_variants_per_shard):
+  return ('_%s_%011d' % (variant.reference_name,
+                         variant.start / number_of_variants_per_shard *
+                         number_of_variants_per_shard),
+          variant)
+
+
+def _compose_multiple_files_to_one(bucket, blob_prefix):
```

Comment:
Very cool to get this working in code!
How long does this take for a whole genome?

Reply:
I tested on a table with 10,000 variants, with 3865 blobs initially generated (I guess the number `1000` max variants is way to small for this case, since there are only 3 variants on average in each file), it takes about 20 minutes to do the compose. I am also interested to know whether there is any parallel solutions for it.